K3s load images

pablochacin opened the pull request:
## What does this PR do?

Adds a function to k3s module for loading images into the k3s cluster.
It also adds helper functions for manipulating images to the docker provider, to facilitate testing.

## Why is it important?

When testing images generated locally (for example, in a CI/CD workflow), it would be convenient not to have to push them to a repository in order to make them available to the k3s test container. Instead, the test should be able tp upload the images to the test container.

## Related issues

- Closes #1621 

<!-- Recommended
## How to test this PR

This PR adds an executable example in modules/k3s/k3s_images_test.go

mdelapenya commented on docker.go:
```diff
@@ -1414,3 +1414,67 @@ func (p *DockerProvider) getDefaultNetwork(ctx context.Context, cli client.APICl
+}
+
+// PullImage pulls image from registry
+func (p *DockerProvider)PullImage(ctx context.Context, image string) error {
```
Could you reuse the code block in https://github.com/testcontainers/testcontainers-go/pull/1622/files#diff-b68159b6168f099a639e8c6f96d75f23333ab0f0505ee22a98268748db17ae4cR1155-R1184 into this PullImage method? It basically does the same but with Retries.

We could remove the `attempToPullImage` method and replace all its usages for this new method (including the old comments).

Thoughts?

pablochacin replied:
It makes sense to have only one method. I will do that.

pablochacin replied:
The only inconvenience I see in renaming `attempToPullImage` is exposing the `types.ImagePullOptions` argument which comes from the docker client package. Maybe `PullImage` should wrap it and pass a default ImagePullOptions?

mdelapenya replied:
Makes sense to me, thanks for considering the downstream usage of the API 🙏

mdelapenya commented on modules/k3s/k3s.go:
```diff
@@ -163,3 +164,21 @@ func unmarshal(bytes []byte) (*KubeConfigValue, error) {
 }
+
+// LoadImages loads images into the k3s container.
+func (c *K3sContainer) LoadImages(ctx context.Context, images string) error {
```
Do we need to validate it's a valid tar file or, on the contrary, it's fine and enough to fail in https://github.com/testcontainers/testcontainers-go/pull/1622/files#diff-85b0f02d7304331291be9d2df1c8a91dc7afd3a7b0bd2d0ffe3afd686a8a99b8R178?

pablochacin replied:
This is validated by docker. I think we only need to propagate docker's errors here.

mdelapenya replied:
We probably need to update the imports for path and filepath :)

mdelapenya reviewed (commented):
I added just two minor comments, although the one regarding the new PullImage method would need your attention.

Other than that, this PR LGTM, thank you very much for your work here 👏

pablochacin commented:
I added some documentation. 
Also, rebased the PR because there were conflicts with main. 
From my side, I think the PR is "complete".
Waiting for your comments.

mdelapenya commented on modules/k3s/k3s_image_test.go:
```diff
@@ -0,0 +1,109 @@
+)
+
+func ExampleLoadImages() {
+	// runK3sContainer {
```
Not needed unless you want to embed this code snippet into our docs site 
```suggestion
```

mdelapenya reviewed (dismissed):
LGTM, added a few comments, all minor.

Thank you so much for your contribution including testable examples in his to load the images into the cluster. Very helpful!

mdelapenya commented on modules/k3s/k3s.go:
```diff
@@ -167,7 +167,7 @@ func unmarshal(bytes []byte) (*KubeConfigValue, error) {
 
 // LoadImages loads images into the k3s container.
 func (c *K3sContainer) LoadImages(ctx context.Context, images string) error {
-	imageFile := path.Base(images)
```
We probably need to update the imports for path and filepath :)

pablochacin replied:
Fixed. Thanks for the catch!

pablochacin commented on modules/k3s/k3s_image_test.go:
```diff
@@ -0,0 +1,107 @@
+package k3s
```
```suggestion
package k3s_test
```

pablochacin commented on modules/k3s/k3s.go:
```diff
@@ -1,9 +1,10 @@
-package k3s
+package k3s_test
```
Sorry, that was a sloppy change, It was intended to be in `k3s_image_test.go`  :disappointed:

mdelapenya commented:
Lint failed: I think you can run `make lint` from the root dir 🤔

mmorel-35 commented on modules/k3s/k3s.go:
```diff
@@ -163,3 +164,21 @@ func unmarshal(bytes []byte) (*KubeConfigValue, error) {
+	}
+
+	return nil
+}
```
You probably need a new line here

pablochacin replied:
Interesting, the linter didn't point out this error :thinking:

mmorel-35 replied:
With gofumpt/goflt you need a new line at the end of the file. But for a reason I ignore gci is the one saying something is wrong . I suspect that gci is trying to apply gofmt plus the import ordering rules

mmorel-35 commented on modules/k3s/k3s_image_test.go:
```diff
@@ -0,0 +1,107 @@
+func ExampleLoadImages() {
+	ctx := context.Background()
+
+	k3sContainer, err := RunContainer(ctx,
```
You probably need to import k3s here as they don't share the same package

mmorel-35 replied:
```suggestion
	k3sContainer, err := k3s.RunContainer(ctx,
```

mmorel-35 commented on modules/k3s/k3s_image_test.go:
```diff
@@ -0,0 +1,107 @@
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
```
```suggestion
	"github.com/testcontainers/testcontainers-go/modules/k3s"
	"github.com/testcontainers/testcontainers-go/wait"
```

pablochacin commented:
Following a conversation with @mdelapenya The `LoadImages` method now receives a list of images and takes the responsibility of saving them in a `.tar` and loading them into the cluster.

As a result of these simplifications, the `LoadImages` runnable example is no longer needed. The `test_k3s_images.go` file now has regular go tests.

mdelapenya commented on docs/modules/k3s.md:
```diff
@@ -79,19 +79,3 @@ to the Kubernetes Rest Client API using a Kubernetes client. It'll be returned i
 <!--codeinclude-->
 [Get KubeConifg](../../modules/k3s/k3s_test.go) inside_block:GetKubeConfig
 <!--/codeinclude-->
-
```
I'm allowing the CI to run to test the PR, if it's green and we add the docs back, then it will LGTM

mdelapenya commented on modules/k3s/k3s_image_test.go:
```diff
@@ -0,0 +1,109 @@
+			t.Fatal(err)
+		}
+
+		if pod.Status.ContainerStatuses[0].State.Waiting.Reason == "ErrImageNeverPull" {
```
The tests say there is a panic here 🤔 Could you take a look 🙏 ?

pablochacin replied:
It should be fixed now. However, the test is taking well over `400s` (> 6 minutes) to complete. I'm concerned this will make it fail due to timeouts.

mdelapenya commented on modules/k3s/k3s_image_test.go:
```diff
@@ -101,8 +101,8 @@ func Test_LoadImages(t *testing.T) {
-
-		if pod.Status.ContainerStatuses[0].State.Waiting.Reason == "ErrImageNeverPull" {
+		waiting := pod.Status.ContainerStatuses[0].State.Waiting
+		if waiting != nil && waiting.Reason == "ErrImageNeverPull" {
```
Could `waiting` be nil? If so, how we are handling it in tests?

pablochacin replied:
If `waiting is `nil` it means the pod is not waiting for the image, so it was properly loaded. We are not interested any other case.

mdelapenya reviewed (approved):
LGTM! Thank you for your time contributing to testcontainers-go! 👏
